flac.go

```diff
@@ -253,7 +275,97 @@ func (stream *Stream) ParseNext() (f *frame.Frame, err error) {
 	return frame.Parse(stream.r)
 }
 
-// Seek is not implement yet.
-func (stream *Stream) Seek(offset int64, whence int) (read int64, err error) {
-	return 0, nil
+// Seek to offset where offset represents the sample number in the flac file
+//
+// TODO: return an error if sample is invalid?  Do nothing?
+//
+// sample could be invalid if:
+// whence == io.SeekEnd && sample is not negative
+// whence == io.SeekStart && sample is negative
+// whence == io.SeekCurrent and sample + current sample < 0 or > stream.Info.NSamples
+func (stream *Stream) Seek(sample int64, whence int) (read int64, err error) {
+	if stream.seekTable == nil {
+		return 0, nil
+	}
+
+	r := stream.r.(io.ReadSeeker)
+
+	var point meta.SeekPoint
+	switch whence {
+	case io.SeekStart:
+		point = stream.searchFromStart(sample)
+	case io.SeekCurrent:
+		o, err := r.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, err
+		}
+		point = stream.searchFromCurrent(sample, o)
+	case io.SeekEnd:
+		point = stream.searchFromEnd(sample)
+	}
+
+	_, err = r.Seek(stream.start+int64(point.Offset), io.SeekStart)
+	return int64(point.SampleNum), err
+}
+
+func (stream *Stream) searchFromCurrent(sample, offset int64) (p meta.SeekPoint) {
+	for _, sp := range stream.seekTable.Points {
+		if int64(sp.Offset) >= offset {
+			return stream.searchFromStart(int64(sp.SampleNum) + sample)
+		}
+	}
+	return p
+}
+
+func (stream *Stream) searchFromEnd(sample int64) (p meta.SeekPoint) {
+	p = stream.seekTable.Points[len(stream.seekTable.Points)-1]
+	return stream.searchFromStart(int64(p.SampleNum) + sample)
+}
+
+func (stream *Stream) searchFromStart(sample int64) (p meta.SeekPoint) {
+	for _, p = range stream.seekTable.Points {
+		if p.SampleNum >= uint64(sample) {
+			return p
+		}
+	}
+	return p
+}
+
+func (stream *Stream) makeSeekTable() (err error) {
+	r := stream.r.(io.ReadSeeker)
+
+	var i int
+	var sampleNum uint64
+	var points []meta.SeekPoint
+	for {
+		f, err := stream.ParseNext()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if i%10 == 0 {
```

Comment:
I made the assumption that the number of seek points can be reduced in order to save memory and increase seek performance (don't have to search through so many points in the seek table) at the cost of the seeks having lower resolution.

Reply:
Yes, that's good!

I don't have a good gut feeling what the perfect resolution is, since this library could possibly handle short sound segments (<2s) to long sound segments (>10h+). This could possibly be a configure option, but let's just go with `%10` for now.

Or do you have any opinions @mewmew ?

Reply:
We could use a sane default (e.g. `10%` as you suggested), and then at a later point introduce a public API for initializing a custom seek table at arbitrary resolution (that API point would essentially set the unexported `seekTable` field).